massa-pos-exports/src/cycle_info.rs

```diff
@@ -0,0 +1,321 @@
+pub struct ProductionStatsDeserializer {
+    address_deserializer: AddressDeserializer,
+    u64_deserializer: U64VarIntDeserializer,
+}
+
+impl Default for ProductionStatsDeserializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ProductionStatsDeserializer {
+    /// Creates a new `ProductionStats` deserializer
+    pub fn new() -> ProductionStatsDeserializer {
+        ProductionStatsDeserializer {
+            address_deserializer: AddressDeserializer::new(),
+            u64_deserializer: U64VarIntDeserializer::new(Included(u64::MIN), Included(u64::MAX)),
+        }
+    }
+}
+
+impl Deserializer<PreHashMap<Address, ProductionStats>> for ProductionStatsDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], PreHashMap<Address, ProductionStats>, E> {
+        context(
+            "Failed ProductionStats deserialization",
+            length_count(
+                context("Failed length deserialization", |input| {
+                    self.u64_deserializer.deserialize(input)
+                }),
+                tuple((
+                    |input| self.address_deserializer.deserialize(input),
+                    |input| self.u64_deserializer.deserialize(input),
+                    |input| self.u64_deserializer.deserialize(input),
+                )),
+            ),
+        )
+        .map(|elements| {
+            elements
+                .into_iter()
+                .map(|(addr, block_success_count, block_failure_count)| {
+                    (
+                        addr,
+                        ProductionStats {
+                            block_success_count,
+                            block_failure_count,
+                        },
+                    )
+                })
+                .collect()
+        })
+        .parse(buffer)
+    }
+}
+
+/// Deserializer for rolls
+pub struct RollsDeserializer {
+    address_deserializer: AddressDeserializer,
+    u64_deserializer: U64VarIntDeserializer,
+}
+
+impl Default for RollsDeserializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RollsDeserializer {
+    /// Creates a new rolls deserializer
+    pub fn new() -> RollsDeserializer {
+        RollsDeserializer {
+            address_deserializer: AddressDeserializer::new(),
+            u64_deserializer: U64VarIntDeserializer::new(Included(u64::MIN), Included(u64::MAX)),
+        }
+    }
+}
+
+impl Deserializer<Vec<(Address, u64)>> for RollsDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], Vec<(Address, u64)>, E> {
+        context(
+            "Failed RollChanges deserialization",
+            length_count(
+                context("Failed length deserialization", |input| {
+                    self.u64_deserializer.deserialize(input)
```

Comment:
A maximum value value should be set for this deserializer to avoid allocating big vector